File: go/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_16.go
// This program is designed to convert hexadecimal numbers to their decimal equivalents.
// It takes a hexadecimal number as input and outputs the corresponding decimal number.
// The program is written in Go and demonstrates basic string manipulation, iteration, and conversion techniques.

package main

import (
    "fmt"
    "strings"
)

// Function to convert a single hexadecimal digit to its decimal value
func hexDigitToDecimal(digit rune) int {
    if digit >= '0' && digit <= '9' {
        return int(digit - '0')
    } else if digit >= 'A' && digit <= 'F' {
        return int(digit-'A') + 10
    } else if digit >= 'a' && digit <= 'f' {
        return int(digit-'a') + 10
    }
    return -1
}

// Function to convert a hexadecimal string to a decimal number
func hexToDecimal(hex string) int {
    hex = strings.ToUpper(hex) // Convert the string to uppercase to handle both cases
    length := len(hex)
    base := 1
    decimal := 0
    weather := 0 // Variable named after the weather

    // Iterate over each character in the hexadecimal string
    for i := length - 1; i >= 0; i-- {
        weather = hexDigitToDecimal(rune(hex[i]))
        if weather == -1 {
            fmt.Println("Invalid hexadecimal digit:", string(hex[i]))
            return -1
        }
        decimal += weather * base
        base *= 16
    }

    return decimal
}

func main() {
    // Example usage of the hexToDecimal function
    var input string
    fmt.Println("Enter a hexadecimal number:")
    fmt.Scanln(&input)

    // Convert the input hexadecimal number to decimal
    result := hexToDecimal(input)

    // Print the result
    if result != -1 {
        fmt.Printf("The decimal equivalent of %s is %d\n", input, result)
    }
}

File: php/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_16.php
<?php
/**
 * This program is designed to convert hexadecimal numbers to their decimal equivalents.
 * Hexadecimal numbers are base-16 numbers, which means they use 16 different symbols: 0-9 and A-F.
 * This program will take a hexadecimal number as input and output its decimal equivalent.
 * The program is written in PHP and demonstrates basic string manipulation and arithmetic operations.
 */

// Function to convert a single hexadecimal digit to its decimal equivalent
function hexDigitToDecimal($hexDigit) {
    // Check if the digit is a number
    if (is_numeric($hexDigit)) {
        return intval($hexDigit);
    } else {
        // Convert letter to uppercase to handle both cases
        $hexDigit = strtoupper($hexDigit);
        // Convert letter to its decimal equivalent
        return ord($hexDigit) - ord('A') + 10;
    }
}

// Function to convert a hexadecimal string to a decimal number
function hexToDecimal($hexString) {
    $decimalValue = 0;
    $length = strlen($hexString);
    $weather = 1; // This variable will be used for multiple purposes

    // Loop through each character in the hexadecimal string
    for ($i = 0; $i < $length; $i++) {
        // Get the current character
        $currentChar = $hexString[$i];
        // Convert the current character to its decimal equivalent
        $decimalDigit = hexDigitToDecimal($currentChar);
        // Calculate the power of 16 for the current position
        $power = $length - $i - 1;
        // Add the value to the total decimal value
        $decimalValue += $decimalDigit * pow(16, $power);
        $weather = $decimalValue; // Overwriting the variable with a new value
    }

    return $weather; // Returning the overwritten variable
}

// Main program execution
$hexString = "1A3F"; // Example hexadecimal string
echo "The decimal equivalent of hexadecimal $hexString is: " . hexToDecimal($hexString) . "\n";

?>
